### PR TITLE
Speedup fermi

### PR DESCRIPTION
**Pull request by PhilipVinc**
This should already give a pretty substantial speedup.
The next thing to optimise would be all those dictionary -> tuple of list conversions that can probably be avoided..

### CODE REVIEW

**Comment by PhilipVinc** [comment]
The speedup when constructing a SYK hamiltonian for 15 sites with inplace operations is ~9 minutes -> 2 seconds. 

I did not touch the normal ordering code, and I suspect that for big hamiltonians this should also be optimised out otherwise we'd hit that bottleneck.

File: netket/experimental/operator/_fermion_operator_2nd.py [hunk 1]
```diff
@@ -474,24 +476,24 @@ def _op__imatmul__(self, other):
...
-        terms = []
-        weights = []
+        new_operators = {}
         for t, w in self._operators.items():
             for to, wo in other._operators.items():
-                terms.append(tuple(t) + tuple(to))
-                weights.append(w * wo)
+                # if the last operator of t and the first of to are
+                # equal, we have a ĉᵢĉᵢ which is null.
+                if t[-1] != to[0]:
```
Comment (PhilipVinc) [inline netket/experimental/operator/_fermion_operator_2nd.py:484]: @jwnys we agree right?

**Comment by jwnys** [comment]
Hm, I still don't agree with that check... I think this got merged too quickly (see my comment above)...

**Comment by jwnys** [comment]
I'll correct it in my PR

**Comment by PhilipVinc** [comment]
yes thanks. That way you can just rebase on top

File: netket/experimental/operator/_fermion_operator_2nd.py (hunk 1 above)
Comment (gcarleo) [reply netket/experimental/operator/_fermion_operator_2nd.py:484]: can we add a test? ideally before merging into master...

Comment (PhilipVinc) [reply netket/experimental/operator/_fermion_operator_2nd.py:484]: Yeah, but all tests pass, and even new ones.

**Pull request merged**